README.md
sptrans
=======

[![Build Status](https://secure.travis-ci.org/diogobaeder/sptrans.png)](http://travis-ci.org/diogobaeder/sptrans)
[![Coverage Status](https://coveralls.io/repos/diogobaeder/sptrans/badge.png)](https://coveralls.io/r/diogobaeder/sptrans)

Python library for the SPTrans API. Read the docs here: http://sptrans.readthedocs.org/


docs/changelog.rst
Changelog
=========

0.1.0
-----

- First version of the library
- All methods from the SPTrans API implemented
- Some documentation is still under development

tests/test_v0.py
# -*- coding: utf-8 -*-
from datetime import date, datetime, time
import json
import os
from unittest import TestCase, skipUnless

from mock import patch
from nose.tools import istest


from . import test_fixtures
from sptrans.v0 import (
    BASE_URL,
    AuthenticationError,
    Client,
    ForecastWithStop,
    ForecastWithStops,
    Lane,
    Route,
    Positions,
    RequestError,
    Stop,
)


TOKEN = os.environ.get('SPTRANS_TOKEN', None)


class ClientTest(TestCase):

    client = None

    def setUp(self):
        self.client = Client()
        self.client._cookies = 'some cookies for later'

    def assert_is_a_generator(self, obj):
        foo = (_ for _ in [])
        generator_type = type(foo)
        self.assertIsInstance(obj, generator_type)

    @istest
    def builds_a_usable_url_from_endpoint_and_parameters(self):
        expected_url = '{}/foo/bar?baz=john+doe'.format(BASE_URL)

        url = self.client._build_url('foo/bar', baz='john doe')

        self.assertEqual(url, expected_url)

    @istest
    @patch('sptrans.v0.requests')
    def gets_content_from_a_certain_endpoint(self, mock_requests):
        url = '{}/foo/bar?baz=joe'.format(BASE_URL)
        content = u'some façade'
        mock_requests.get.return_value.content = content.encode('latin1')

        content = self.client._get_content('foo/bar', baz='joe')

        self.assertEqual(content, content)
        mock_requests.get.assert_called_once_with(url, cookies=self.client._cookies)

    @istest
    @patch('sptrans.v0.requests')
    def authenticates_the_user(self, mock_requests):
        token = 'some token'
        client = Client()
        mock_requests.post.return_value.content = b'true'

        client.authenticate(token)

        url = self.client._build_url('Login/Autenticar', token=token)
        mock_requests.post.assert_called_once_with(url)

    @istest
    @patch('sptrans.v0.requests')
    def cannot_authenticate_the_user_if_token_is_invalid(self, mock_requests):
        token = 'some wrong token'
        client = Client()
        mock_requests.post.return_value.content = b'false'

        self.assertRaises(AuthenticationError, client.authenticate, token)

    @istest
    @patch('sptrans.v0.requests')
    def keeps_cookies_for_later(self, mock_requests):
        token = 'some token'
        client = Client()

        class response:
            cookies = 'some cookies'
            content = b'true'
        mock_requests.post.return_value = response

        client.authenticate(token)

        self.assertEqual(client._cookies, response.cookies)

    @istest
    @patch('sptrans.v0.requests')
    def searches_routes(self, mock_requests):
        keywords = 'my search'

        mock_requests.get.return_value.content = test_fixtures.ROUTE_SEARCH

        routes = list(self.client.search_routes(keywords))

        expected_routes = [Route.from_dict(route_dict)
                           for route_dict in json.loads(test_fixtures.ROUTE_SEARCH.decode('latin1'))]
        url = self.client._build_url('Linha/Buscar', termosBusca=keywords)
        self.assertEqual(routes, expected_routes)
        mock_requests.get.assert_called_once_with(url, cookies=self.client._cookies)

    @istest
    @patch('sptrans.v0.requests')
    def searches_routes_returns_generator(self, mock_requests):
        keywords = 'my search'

        mock_requests.get.return_value.content = test_fixtures.ROUTE_SEARCH

        routes = self.client.search_routes(keywords)

        self.assert_is_a_generator(routes)

    @istest
    @patch('sptrans.v0.requests')
    def searches_stops(self, mock_requests):
        keywords = 'my search'

        mock_requests.get.return_value.content = test_fixtures.STOP_SEARCH

        stops = list(self.client.search_stops(keywords))

        expected_stops = [Stop.from_dict(stop_dict)
                          for stop_dict in json.loads(test_fixtures.STOP_SEARCH.decode('latin1'))]
        url = self.client._build_url('Parada/Buscar', termosBusca=keywords)
        self.assertEqual(stops, expected_stops)
        mock_requests.get.assert_called_once_with(url, cookies=self.client._cookies)

    @istest
    @patch('sptrans.v0.requests')
    def searches_stops_by_route(self, mock_requests):
        code = '1234'

        mock_requests.get.return_value.content = test_fixtures.STOP_SEARCH_BY_ROUTE

        stops = list(self.client.search_stops_by_route(code))

        expected_stops = [Stop.from_dict(stop_dict)
                          for stop_dict in json.loads(test_fixtures.STOP_SEARCH_BY_ROUTE.decode('latin1'))]
        url = self.client._build_url('Parada/BuscarParadasPorLinha', codigoLinha=code)
        self.assertEqual(stops, expected_stops)
        mock_requests.get.assert_called_once_with(url, cookies=self.client._cookies)

    @istest
    @patch('sptrans.v0.requests')
    def searches_stops_by_lane(self, mock_requests):
        code = '1234'

        mock_requests.get.return_value.content = test_fixtures.STOP_SEARCH_BY_LANE

        stops = list(self.client.search_stops_by_lane(code))

        expected_stops = [Stop.from_dict(stop_dict)
                          for stop_dict in json.loads(test_fixtures.STOP_SEARCH_BY_LANE.decode('latin1'))]
        url = self.client._build_url('Parada/BuscarParadasPorCorredor', codigoCorredor=code)
        self.assertEqual(stops, expected_stops)
        mock_requests.get.assert_called_once_with(url, cookies=self.client._cookies)

    @istest
    @patch('sptrans.v0.requests')
    def lists_lanes(self, mock_requests):
        mock_requests.get.return_value.content = test_fixtures.LANES

        lanes = list(self.client.list_lanes())

        expected_lanes = [Lane.from_dict(lane_dict)
                          for lane_dict in json.loads(test_fixtures.LANES.decode('latin1'))]
        url = self.client._build_url('Corredor')
        self.assertEqual(lanes, expected_lanes)
        mock_requests.get.assert_called_once_with(url, cookies=self.client._cookies)

    @istest
    @patch('sptrans.v0.requests')
    def gets_positions(self, mock_requests):
        fixture = test_fixtures.VEHICLE_POSITIONS
        code = '1234'

        mock_requests.get.return_value.content = fixture

        positions = self.client.get_positions(code)

        expected_positions = Positions.from_dict(json.loads(fixture.decode('latin1')))
        url = self.client._build_url('Posicao', codigoLinha=code)
        self.assertEqual(positions, expected_positions)
        mock_requests.get.assert_called_once_with(url, cookies=self.client._cookies)

    @istest
    @patch('sptrans.v0.requests')
    def gets_forecast_for_route_and_stop(self, mock_requests):
        fixture = test_fixtures.FORECAST_FOR_ROUTE_AND_STOP
        stop_code = '1234'
        route_code = '2345'

        mock_requests.get.return_value.content = fixture

        forecast = self.client.get_forecast(stop_code=stop_code, route_code=route_code)

        expected_forecast = ForecastWithStop.from_dict(json.loads(fixture.decode('latin1')))
        url = self.client._build_url('Previsao', codigoParada=stop_code, codigoLinha=route_code)
        self.assertEqual(forecast, expected_forecast)
        mock_requests.get.assert_called_once_with(url, cookies=self.client._cookies)

    @istest
    @patch('sptrans.v0.requests')
    def gets_forecast_for_route(self, mock_requests):
        fixture = test_fixtures.FORECAST_FOR_ROUTE
        route_code = '2345'

        mock_requests.get.return_value.content = fixture

        forecast = self.client.get_forecast(route_code=route_code)

        expected_forecast = ForecastWithStops.from_dict(json.loads(fixture.decode('latin1')))
        url = self.client._build_url('Previsao/Linha', codigoLinha=route_code)
        self.assertEqual(forecast, expected_forecast)
        mock_requests.get.assert_called_once_with(url, cookies=self.client._cookies)

    @istest
    @patch('sptrans.v0.requests')
    def gets_forecast_for_stop(self, mock_requests):
        fixture = test_fixtures.FORECAST_FOR_STOP
        stop_code = '1234'

        mock_requests.get.return_value.content = fixture

        forecast = self.client.get_forecast(stop_code=stop_code)

        expected_forecast = ForecastWithStop.from_dict(json.loads(fixture.decode('latin1')))
        url = self.client._build_url('Previsao/Parada', codigoParada=stop_code)
        self.assertEqual(forecast, expected_forecast)
        mock_requests.get.assert_called_once_with(url, cookies=self.client._cookies)

    @istest
    @patch('sptrans.v0.requests')
    def raises_request_error_if_not_authenticated(self, mock_requests):
        fixture = test_fixtures.MESSAGE_ERROR
        mock_requests.get.return_value.content = fixture

        expected_message = json.loads(fixture.decode('latin1'))[u'Message']

        self.assertRaisesRegexp(RequestError, expected_message, self.client._get_json, 'Some/Endpoint')


@skipUnless(TOKEN, 'Please provide an SPTRANS_TOKEN env variable')
class ClientFunctionalTest(TestCase):
    def setUp(self):
        self.client = Client()
        self.client.authenticate(TOKEN)

    @istest
    def searches_routes(self):
        keywords = 'lapa'
        routes = list(self.client.search_routes(keywords))

        self.assertGreater(len(routes), 0)
        for route in routes:
            main = route.main_to_sec.lower()
            sec = route.sec_to_main.lower()
            self.assertTrue((keywords in main) or (keywords in sec))


class RouteTest(TestCase):

    @istest
    def converts_a_dict_to_a_route(self):
        route_dict = json.loads(test_fixtures.ROUTE_SEARCH.decode('latin1'))[0]

        route = Route.from_dict(route_dict)

        self.assertEqual(route.code, 1273)
        self.assertEqual(route.circular, False)
        self.assertEqual(route.sign, '8000')
        self.assertEqual(route.direction, 1)
        self.assertEqual(route.type, 10)
        self.assertEqual(route.main_to_sec, 'PCA.RAMOS DE AZEVEDO')
        self.assertEqual(route.sec_to_main, 'TERMINAL LAPA')
        self.assertEqual(route.info, None)


class StopTest(TestCase):

    @istest
    def converts_a_dict_to_a_stop(self):
        stop_dict = json.loads(test_fixtures.STOP_SEARCH.decode('latin1'))[0]

        stop = Stop.from_dict(stop_dict)

        self.assertEqual(stop.code, 340015329)
        self.assertEqual(stop.name, "AFONSO BRAZ B/C1")
        self.assertEqual(stop.address, 'R ARMINDA/ R BALTHAZAR DA VEIGA')
        self.assertEqual(stop.latitude, -23.592938)
        self.assertEqual(stop.longitude, -46.672727)


class LaneTest(TestCase):

    @istest
    def converts_a_dict_to_a_lane(self):
        lane_dict = {
            'CodCorredor': 8,
            'CodCot': 0,
            'Nome': 'Campo Limpo',
        }

        lane = Lane.from_dict(lane_dict)

        self.assertEqual(lane.code, 8)
        self.assertEqual(lane.cot, 0)
        self.assertEqual(lane.name, 'Campo Limpo')


class PositionsTest(TestCase):

    @istest
    def converts_a_dict_to_a_positions_object_with_vehicles(self):
        positions_dict = json.loads(test_fixtures.VEHICLE_POSITIONS.decode('latin1'))
        today = date.today()
        hour = time(hour=22, minute=57)

        positions = Positions.from_dict(positions_dict)

        self.assertEqual(positions.time, datetime.combine(today, hour))

        self.assertEqual(positions.vehicles[0].prefix, '11433')
        self.assertEqual(positions.vehicles[0].accessible, False)
        self.assertEqual(positions.vehicles[0].latitude, -23.540150375000003)
        self.assertEqual(positions.vehicles[0].longitude, -46.64414075)

        self.assertEqual(positions.vehicles[1].prefix, '12132')
        self.assertEqual(positions.vehicles[1].accessible, False)
        self.assertEqual(positions.vehicles[1].latitude, -23.5200315)
        self.assertEqual(positions.vehicles[1].longitude, -46.699387)


class ForecastWithStopTest(TestCase):

    @istest
    def converts_a_dict_to_an_forecast_object_with_stop_routes_and_vehicles(self):
        forecast_dict = json.loads(test_fixtures.FORECAST_FOR_ROUTE_AND_STOP.decode('latin1'))
        today = date.today()

        forecast = ForecastWithStop.from_dict(forecast_dict)

        self.assertEqual(forecast.time, datetime.combine(today, time(hour=23, minute=9)))

        self.assertEqual(forecast.stop.code, 4200953)
        self.assertEqual(forecast.stop.name, 'PARADA ROBERTO SELMI DEI B/C')
        self.assertEqual(forecast.stop.latitude, -23.675901)
        self.assertEqual(forecast.stop.longitude, -46.752812)

        self.assertEqual(forecast.stop.routes[0].sign, '7021-10')
        self.assertEqual(forecast.stop.routes[0].code, 1989)
        self.assertEqual(forecast.stop.routes[0].direction, 1)
        self.assertEqual(forecast.stop.routes[0].main_to_sec, u'TERM. JOÃO DIAS')
        self.assertEqual(forecast.stop.routes[0].sec_to_main, u'JD. MARACÁ')
        self.assertEqual(forecast.stop.routes[0].quantity, 1)

        self.assertEqual(forecast.stop.routes[0].vehicles[0].prefix, '74558')
        self.assertEqual(forecast.stop.routes[0].vehicles[0].arriving_at, datetime.combine(today, time(hour=23, minute=11)))
        self.assertEqual(forecast.stop.routes[0].vehicles[0].accessible, True)
        self.assertEqual(forecast.stop.routes[0].vehicles[0].latitude, -23.67603)
        self.assertEqual(forecast.stop.routes[0].vehicles[0].longitude, -46.75891166666667)


class ForecastWithStopsTest(TestCase):

    @istest
    def converts_a_dict_to_an_forecast_object_with_stops_from_route(self):
        forecast_dict = json.loads(test_fixtures.FORECAST_FOR_ROUTE.decode('latin1'))
        today = date.today()

        forecast = ForecastWithStops.from_dict(forecast_dict)

        self.assertEqual(forecast.time, datetime.combine(today, time(hour=23, minute=18)))

        self.assertEqual(forecast.stops[0].code, 700016623)
        self.assertEqual(forecast.stops[0].name, 'ANA CINTRA B/C')
        self.assertEqual(forecast.stops[0].latitude, -23.538763)
        self.assertEqual(forecast.stops[0].longitude, -46.646925)

        self.assertEqual(forecast.stops[0].vehicles[0].prefix, '11436')
        self.assertEqual(forecast.stops[0].vehicles[0].arriving_at, datetime.combine(today, time(hour=23, minute=26)))
        self.assertEqual(forecast.stops[0].vehicles[0].accessible, False)
        self.assertEqual(forecast.stops[0].vehicles[0].latitude, -23.528119999999998)
        self.assertEqual(forecast.stops[0].vehicles[0].longitude, -46.670674999999996)


sptrans/v0.py
"""Module for the v0 version of the `SPTrans API <http://www.sptrans.com.br/desenvolvedores/APIOlhoVivo/Documentacao.aspx?1>`_.

The first thing you have to do, in order to use it, is to instantiate a client and authenticate to the API:
::

    from sptrans.v0 import Client


    client = Client()
    client.authenticate('this is my token')

Then you can use the other methods to grab data from the API.
"""

from collections import namedtuple
from datetime import date, datetime, time
import json
try:
    from urllib import urlencode
except ImportError:  # pragma: no cover
    from urllib.parse import urlencode

import requests


BASE_URL = 'http://api.olhovivo.sptrans.com.br/v0'


class RequestError(Exception):
    """Raised when the request failes to be accomplished.

    Normally this is due to the client not being authenticated anymore.
    In this case, just authenticate again, and it should be back at work.
    """


class AuthenticationError(Exception):
    """Raised when the authentication fails - for example, with a wrong token -."""


class TupleMapMixin(object):
    MAPPING = {}

    @classmethod
    def from_dict(cls, result_dict):
        kwargs = {}
        for key, value in cls.MAPPING.items():
            if isinstance(value, str):
                kwargs[key] = result_dict[value]
            else:
                kwargs[key] = value.resolve(result_dict)
        return cls(**kwargs)


def build_tuple_class(name, mapping):
    base_classes = (namedtuple(name, mapping.keys()), TupleMapMixin)
    return type(name, base_classes, {'MAPPING': mapping})


def time_string_to_datetime(time_string):
    hour_parts = time_string.split(':')
    hour, minute = [int(part) for part in hour_parts]
    return datetime.combine(date.today(), time(hour=hour, minute=minute))


class TimeField(object):

    def __init__(self, field):
        self.field = field

    def resolve(self, result_dict):
        return time_string_to_datetime(result_dict[self.field])


class TupleField(object):
    def __init__(self, field, tuple_class):
        self.field = field
        self.tuple_class = tuple_class

    def resolve(self, result_dict):
        return self.tuple_class.from_dict(result_dict[self.field])


class TupleListField(object):
    def __init__(self, field, tuple_class):
        self.field = field
        self.tuple_class = tuple_class

    def resolve(self, result_dict):
        return [self.tuple_class.from_dict(internal_dict)
                for internal_dict in result_dict[self.field]]


Route = build_tuple_class('Route', {
    'code': 'CodigoLinha',
    'circular': 'Circular',
    'sign': 'Letreiro',
    'direction': 'Sentido',
    'type': 'Tipo',
    'main_to_sec': 'DenominacaoTPTS',
    'sec_to_main': 'DenominacaoTSTP',
    'info': 'Informacoes',
})
"""A namedtuple representing a route.

:var code: (:class:`int`) The route code.
:var circular: (:class:`bool`) Wether it's a circular route or not (without a secondary terminal).
:var sign: (:class:`str`) The first part of the route sign.
:var direction: (:class:`int`) The route direction. "1" means "main to secondary terminal", "2" means "secondary to main terminal".
:var type: (:class:`int`) The route type.
:var main_to_sec: (:class:`str`) The name of the route when moving from the main terminal to the second one.
:var sec_to_main: (:class:`str`) The name of the route when moving from the second terminal to the main one.
:var info: (:class:`str`) Extra information about the route.
"""
Stop = build_tuple_class('Stop', {
    'code': 'CodigoParada',
    'name': 'Nome',
    'address': 'Endereco',
    'latitude': 'Latitude',
    'longitude': 'Longitude',
})
"""A namedtuple representing a bus stop.

:var code: (:class:`int`) The stop code.
:var name: (:class:`str`) The stop name.
:var address: (:class:`str`) The stop address.
:var latitude: (:class:`float`) The stop latitude.
:var longitude: (:class:`float`) The stop longitude.
"""
Lane = build_tuple_class('Lane', {
    'code': 'CodCorredor',
    'cot': 'CodCot',
    'name': 'Nome',
})
"""A namedtuple representing a bus lane.

:var code: (:class:`int`) The lane code.
:var cot: (:class:`int`) The lane "cot" (?).
:var name: (:class:`str`) The lane name.
"""
Vehicle = build_tuple_class('Vehicle', {
    'prefix': 'p',
    'accessible': 'a',
    'latitude': 'py',
    'longitude': 'px',
})
"""A namedtuple representing a vehicle (bus) with its position.

:var prefix: (:class:`str`) The vehicle prefix painted in the bus.
:var accessible: (:class:`bool`) Wether the vehicle is accessible or not.
:var latitude: (:class:`float`) The vehicle latitude.
:var longitude: (:class:`float`) The vehicle longitude.
"""
VehicleForecast = build_tuple_class('VehicleForecast', {
    'prefix': 'p',
    'accessible': 'a',
    'arriving_at': TimeField('t'),
    'latitude': 'py',
    'longitude': 'px',
})
"""A namedtuple representing a vehicle (bus) with its position and forecast to arrive at a certain stop.

:var prefix: (:class:`str`) The vehicle prefix painted in the bus.
:var accessible: (:class:`bool`) Wether the vehicle is accessible or not.
:var arriving_at: (:class:`datetime.datetime`) The time that the vehicle is expected to arrive.
:var latitude: (:class:`float`) The vehicle latitude.
:var longitude: (:class:`float`) The vehicle longitude.
"""
Positions = build_tuple_class('Positions', {
    'time': TimeField('hr'),
    'vehicles': TupleListField('vs', Vehicle),
})
"""A namedtuple representing a sequence of vehicles positions, with the time when the information was retrieved.

:var time: (:class:`datetime.datetime`) The time when the information was retrieved.
:var vehicles: (:class:`list`) The list of :class:`vehicles <Vehicle>`.
"""
RouteWithVehicles = build_tuple_class('RouteWithVehicles', {
    'sign': 'c',
    'code': 'cl',
    'direction': 'sl',
    'main_to_sec': 'lt0',
    'sec_to_main': 'lt1',
    'quantity': 'qv',
    'vehicles': TupleListField('vs', VehicleForecast),
})
"""A namedtuple representing a route with a sequence of vehicles with their current positions.

:var sign: (:class:`str`) The first part of the route sign.
:var code: (:class:`int`) The route code.
:var direction: (:class:`int`) The route direction. "1" means "main to secondary terminal", "2" means "secondary to main terminal".
:var main_to_sec: (:class:`str`) The name of the route when moving from the main terminal to the second one.
:var sec_to_main: (:class:`str`) The name of the route when moving from the second terminal to the main one.
:var quantity: (:class:`int`) The quantity of vehicles.
:var vehicles: (:class:`list`) The list of :class:`vehicles <Vehicle>`.
"""
StopWithRoutes = build_tuple_class('StopWithRoutes', {
    'code': 'cp',
    'name': 'np',
    'latitude': 'py',
    'longitude': 'px',
    'routes': TupleListField('l', RouteWithVehicles),
})
"""A namedtuple representing a bus stop with a list of routes that pass through this stop.

:var code: (:class:`int`) The stop code.
:var name: (:class:`str`) The stop name.
:var latitude: (:class:`float`) The stop latitude.
:var longitude: (:class:`float`) The stop longitude.
:var routes: (:class:`list`) The list of :class:`routes <Route>` that pass through this stop.
"""
StopWithVehicles = build_tuple_class('StopWithVehicles', {
    'code': 'cp',
    'name': 'np',
    'latitude': 'py',
    'longitude': 'px',
    'vehicles': TupleListField('vs', VehicleForecast),
})
"""A namedtuple representing a bus stop with a list of vehicles that pass through this stop.

:var code: (:class:`int`) The stop code.
:var name: (:class:`str`) The stop name.
:var latitude: (:class:`float`) The stop latitude.
:var longitude: (:class:`float`) The stop longitude.
:var vehicles: (:class:`list`) The list of :class:`vehicles <Vehicle>`.
"""
ForecastWithStop = build_tuple_class('ForecastWithStop', {
    'time': TimeField('hr'),
    'stop': TupleField('p', StopWithRoutes),
})
"""A namedtuple representing a bus stop forecast with routes and the time when the information was retrieved.

:var time: (:class:`datetime.datetime`) The time when the information was retrieved.
:var stop: (:class:`StopWithRoutes`) The bus stop with :class:`routes <Route>`.
"""
ForecastWithStops = build_tuple_class('ForecastWithStops', {
    'time': TimeField('hr'),
    'stops': TupleListField('ps', StopWithVehicles),
})
"""A namedtuple representing a list of bus stops forecast with vehicles and the time when the information was retrieved.

:var time: (:class:`datetime.datetime`) The time when the information was retrieved.
:var stops: (:class:`list` of :class:`StopWithVehicles`) The bus stops.
"""


class Client(object):
    """Main client class.

    .. warning:: Any method (except :meth:`authenticate`) may raise :class:`RequestError` if the client is not authenticated anymore.
       So keep an eye on the authentication state.

    Example:
    ::

        from sptrans.v0 import Client


        client = Client()
        client.authenticate('this is my token')

    """
    _cookies = None

    def _build_url(self, endpoint, **kwargs):
        query_string = urlencode(kwargs)
        return '{}/{}?{}'.format(BASE_URL, endpoint, query_string)

    def _get_content(self, endpoint, **kwargs):
        url = self._build_url(endpoint, **kwargs)
        response = requests.get(url, cookies=self._cookies)
        return response.content.decode('latin1')

    def _get_json(self, endpoint, **kwargs):
        content = self._get_content(endpoint, **kwargs)
        result = json.loads(content)
        if isinstance(result, dict) and tuple(result.keys()) == (u'Message', ):
            raise RequestError(result[u'Message'])
        return result

    def authenticate(self, token):
        """Authenticates to the webservice.

        :param token: The API token string.
        :type token: :class:`str`
        :raises: :class:`AuthenticationError` when there's an error during authentication.
        """
        url = self._build_url('Login/Autenticar', token=token)
        response = requests.post(url)
        result = json.loads(response.content.decode('latin1'))
        if not result:
            raise AuthenticationError('Cannot authenticate with token "{}"'.format(token))
        self._cookies = response.cookies

    def search_routes(self, keywords):
        """Searches for routes that match the provided keywords.

        :param keywords: The keywords, in a single string, to use for matching.
        :type keywords: :class:`str`
        :return: A generator that yields :class:`Route` objects.

        Example:
        ::

            from sptrans.v0 import Client


            client = Client()
            client.authenticate('this is my token')
            for route in client.search_routes('butanta'):
                print(route.code, route.sign)

        """
        result_list = self._get_json('Linha/Buscar', termosBusca=keywords)
        for result_dict in result_list:
            yield Route.from_dict(result_dict)

    def search_stops(self, keywords):
        """Searches for bus stops that match the provided keywords.

        :param keywords: The keywords, in a single string, to use for matching.
        :type keywords: :class:`str`
        :return: A generator that yields :class:`Stop` objects.

        Example:
        ::

            from sptrans.v0 import Client


            client = Client()
            client.authenticate('this is my token')
            for stop in client.search_stops('butanta'):
                print(stop.code, stop.name)
        """
        result_list = self._get_json('Parada/Buscar', termosBusca=keywords)
        for result_dict in result_list:
            yield Stop.from_dict(result_dict)

    def search_stops_by_route(self, code):
        """Searches for bus stops that are passed by the route specified by its code.

        :param code: The route code to use for matching.
        :type code: :class:`int`
        :return: A generator that yields :class:`Stop` objects.

        Example:
        ::

            from sptrans.v0 import Client


            client = Client()
            client.authenticate('this is my token')
            for stop in client.search_stops_by_route(1234):
                print(stop.code, stop.name)
        """
        result_list = self._get_json('Parada/BuscarParadasPorLinha', codigoLinha=code)
        for result_dict in result_list:
            yield Stop.from_dict(result_dict)

    def search_stops_by_lane(self, code):
        """Searches for bus stops that are contained in a lane specified by its code.

        :param code: The lane code to use for matching.
        :type code: :class:`int`
        :return: A generator that yields :class:`Stop` objects.

        Example:
        ::

            from sptrans.v0 import Client


            client = Client()
            client.authenticate('this is my token')
            for stop in client.search_stops_by_lane(1234):
                print(stop.code, stop.name)
        """
        result_list = self._get_json('Parada/BuscarParadasPorCorredor', codigoCorredor=code)
        for result_dict in result_list:
            yield Stop.from_dict(result_dict)

    def list_lanes(self):
        """Lists all the bus lanes in the city.

        :return: A generator that yields :class:`Lane` objects.

        Example:
        ::

            from sptrans.v0 import Client


            client = Client()
            client.authenticate('this is my token')
            for lane in client.list_lanes():
                print(lane.code, lane.name)
        """
        result_list = self._get_json('Corredor')
        for result_dict in result_list:
            yield Lane.from_dict(result_dict)

    def get_positions(self, code):
        """Gets the vehicles with their current positions, provided a route code.

        :param code: The route code to use for matching.
        :type code: :class:`int`
        :return: A single :class:`Positions` object.

        Example:
        ::

            from sptrans.v0 import Client


            client = Client()
            client.authenticate('this is my token')

            positions = client.get_positions(1234)
            print(positions.time)

            for vehicle in positions.vehicles:
                print(vehicle.prefix)
        """
        result_dict = self._get_json('Posicao', codigoLinha=code)
        return Positions.from_dict(result_dict)

    def get_forecast(self, stop_code=None, route_code=None):
        """Gets the arrival forecast, provided a route code or a stop code or both.

        You must provide at least one of the parameters.
        If you provide only a `stop_code`, it will return a forecast for all routes that attend a certain stop.
        If you provide only a `route_code`, it will return a forecast for all stops that a certain route attends to.
        If you provide both, it will return a forecast for the specific route attending to the specific stop provided.

        :param stop_code: The stop code to use for matching.
        :type stop_code: :class:`int`
        :param route_code: The stop code to use for matching.
        :type route_code: :class:`int`
        :return: A single :class:`ForecastWithStop` object, when passing only `stop_code` or both.
        :return: A single :class:`ForecastWithStops` object, when passing only `route_code`.

        Example:
        ::

            from sptrans.v0 import Client


            client = Client()
            client.authenticate('this is my token')

            forecast = client.get_forecast(stop_code=1234)
            for route in forecast.stop.routes:
                for vehicle in route.vehicles:
                    print(vehicle.prefix)

            forecast = client.get_forecast(stop_code=1234, route_code=2345)
            for route in forecast.stop.routes:
                for vehicle in route.vehicles:
                    print(vehicle.prefix)

            forecast = client.get_forecast(route_code=2345)
            for stop in forecast.stops:
                for vehicle in stop.vehicles:
                    print(vehicle.prefix)
        """
        if stop_code is None:
            result_dict = self._get_json('Previsao/Linha', codigoLinha=route_code)
            return ForecastWithStops.from_dict(result_dict)

        if route_code is None:
            result_dict = self._get_json('Previsao/Parada', codigoParada=stop_code)
        else:
            result_dict = self._get_json('Previsao', codigoParada=stop_code, codigoLinha=route_code)
        return ForecastWithStop.from_dict(result_dict)
